Filter entities from logbook

wokar opened the pull request:
**Description:**
Added the ability to exclude logbook entries for certain entities or domains.
- exclude logbook messages from hidden entities
- exclude logbook messages for entities or domains configured in the logbook platform

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation:** home-assistant/home-assistant.io#963

**Example entry for `configuration.yaml`:**

``` yaml
logbook:
  exclude:
    entities:
      - sensor.motion_basement_temperature_5
      - sensor.last_boot
      - sensor.date
    domains:
      - sun
      - weblink
```

**Checklist:**
- [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] Tests have been added to verify that the new code works.

wokar committed:
o added ability to exclude entities or domains from logbook
o exclude hidden entities

wokar committed:
fixed remaned configuration key

tchellomello commented:
Nice job @wokar !! I was working on a similar thing here as well!!  We might need to do this the same for history  component. Are you also working on it?

wokar commented:
Not yet, but my plan was to improve the logbook and the history in similar ways. I started with the logbook, because it looked easier to me. After this PR is merged I wanted to start working on the history.
In my setup, with a lot of zwave sensors, the history is really slow and shows a lot of information I don't really care about. IMHO filtering will be a huge improvement of the history component.

If you are already working on the history - just go ahead. We should just make sure that the configuration is similar or even the same.

arsaboo commented:
Looking forward to this....there is indeed lot of noise in both the history and logbook. This should make them faster and may also help reduce the read/write to the SD card.

fanaticDavid commented:
Would it be easy to also allow users to specify which entities to **include** rather than exclude? In other words: provide a way to either blacklist or whitelist entities for the logbook/history.
E.g.: if a user has 100 entities but only wishes to track 10, it would be easier to specify the 10 entities to track than the 90 entities not to track.

balloob commented on homeassistant/components/logbook.py:
```diff
@@ -97,7 +116,7 @@ def get(self, request, datetime=None):
         events = recorder.execute(query)
 
-        return self.json(humanify(events))
+        return self.json(humanify(events, self.config))
```
Let's filter the events before they get passed to humanify, to separate concerns 👍

balloob commented on homeassistant/components/logbook.py:
```diff
@@ -239,6 +264,24 @@ def humanify(events):
+            CONF_ENTITIES in exclude else []
+        exclude_platforms = exclude.get(CONF_PLATFORMS) if \
+            CONF_PLATFORMS in exclude else []
+    customize = config[HA_DOMAIN].get(CONF_CUSTOMIZE)
```
Instead of looking at `customize`, look for the hidden attribute on the state change events.

balloob commented on homeassistant/components/logbook.py:
```diff
@@ -239,6 +264,24 @@ def humanify(events):
+    exclude_platforms = []
+    exclude = config[DOMAIN].get(CONF_EXCLUDE) if DOMAIN in config else None
+    if exclude:
+        exclude_entities = exclude.get(CONF_ENTITIES) if \
```
`CONF_ENTITIES` defaults to an empty list so will always be there

balloob commented on homeassistant/components/logbook.py:
```diff
@@ -239,6 +264,24 @@ def humanify(events):
+    if exclude:
+        exclude_entities = exclude.get(CONF_ENTITIES) if \
+            CONF_ENTITIES in exclude else []
+        exclude_platforms = exclude.get(CONF_PLATFORMS) if \
```
`CONF_PLATFORMS` defaults to an empty list so will always be there

balloob reviewed (changes requested):
Just a few tweaks needed, looks good.

wokar committed:
- filter the events before they get passed to humanify, to separate concerns
- instead of looking at customize, look for the hidden attribute on the state change events
- access to configuration defaults to an empty list - no need to check

wokar commented:
@balloob thanks for the feedback - changes are checked in.

@fanaticDavid to add an `include` configuration should be no problem. If time permitts I will add this in a future PR.

balloob commented on homeassistant/components/logbook.py:
```diff
@@ -239,6 +259,38 @@ def humanify(events):
+    excluded_domains = []
+    exclude = config[DOMAIN].get(CONF_EXCLUDE)
+    if exclude:
+        excluded_entities = exclude.get(CONF_ENTITIES, [])
```
You've already specified a default value in `CONFIG_SCHEMA`, so no need to specify one here. You can just use `exclude[CONF_ENTITIES]`

balloob replied:
I see now that this check is needed because your tests contain the wrong config. Best to run your `EMPTY_CONFIG` through `CONFIG_SCHEMA` so defaults get set (and it keeps working in the future).

wokar replied:
Thanks for the tip - updated the tests...

wokar committed:
- filter only events of type EVENT_STATE_CHANGED
- improve config handling

balloob commented:
Nice! 🐬

wokar committed:
added unit tests to cover all filter cases and logbook message creation
